pkg/operator/ceph/file/mds/mds.go

```diff
@@ -119,84 +132,166 @@ func (c *Cluster) Start() error {
 		if err != nil {
-			return errors.Wrapf(err, "failed to set owner reference for mds deployment %q", d.Name)
+			return "", errors.Wrapf(err, "failed to get existing mds deployment %s for update", d.Name)
 		}
+	}
 
-		// Set the deployment hash as an annotation
-		err = patch.DefaultAnnotator.SetLastAppliedAnnotation(d)
-		if err != nil {
-			return errors.Wrapf(err, "failed to set annotation for deployment %q", d.Name)
+	if createErr != nil && kerrors.IsAlreadyExists(createErr) {
+		if err = UpdateDeploymentAndWait(c.context, c.clusterInfo, d, config.MdsType, daemonLetterID, c.clusterSpec.SkipUpgradeChecks, c.clusterSpec.ContinueUpgradeAfterChecksEvenIfNotHealthy); err != nil {
+			return "", errors.Wrapf(err, "failed to update mds deployment %s", d.Name)
 		}
+	}
+	return d.GetName(), nil
+}
 
-		_, createErr := c.context.Clientset.AppsV1().Deployments(c.fs.Namespace).Create(ctx, d, metav1.CreateOptions{})
-		if createErr != nil {
-			if !kerrors.IsAlreadyExists(createErr) {
-				return errors.Wrapf(createErr, "failed to create mds deployment %s", mdsConfig.ResourceName)
-			}
-			logger.Infof("deployment for mds %s already exists. updating if needed", mdsConfig.ResourceName)
-			_, err = c.context.Clientset.AppsV1().Deployments(c.fs.Namespace).Get(ctx, d.Name, metav1.GetOptions{})
-			if err != nil {
-				return errors.Wrapf(err, "failed to get existing mds deployment %s for update", d.Name)
-			}
-		}
+// isUpgradeMDS determine if mds version inferior than image
+func (c *Cluster) isUpgradeMDS() (bool, error) {
 
-		if createErr != nil && kerrors.IsAlreadyExists(createErr) {
-			if err = UpdateDeploymentAndWait(c.context, c.clusterInfo, d, config.MdsType, daemonLetterID, c.clusterSpec.SkipUpgradeChecks, c.clusterSpec.ContinueUpgradeAfterChecksEvenIfNotHealthy); err != nil {
-				return errors.Wrapf(err, "failed to update mds deployment %s", d.Name)
-			}
+	allVersions, err := client.GetAllCephDaemonVersions(c.context, c.clusterInfo)
+
+	if err != nil {
+		return false, err
+	}
+	monVersion := c.clusterInfo.CephVersion
+	for key := range allVersions.Mds {
+		currentVersion, err := cephver.ExtractCephVersion(key)
+		if err != nil {
+			return false, err
+		}
+		if monVersion.Major > currentVersion.Major {
+			logger.Debugf("current ceph version is %v and mon version is %v", currentVersion, monVersion)
+			return true, err
 		}
-		desiredDeployments[d.GetName()] = true // add deployment name to improvised set
+	}
+
+	return false, nil
+}
+
+func (c *Cluster) upgradeMDS() error {
 
+	// 1. set allow_standby_replay to false
+	if err := client.AllowStandbyReplay(c.context, c.clusterInfo, c.fs.Name, false); err != nil {
+		return errors.Wrapf(err, "failed to setting allow_standby_replay to false")
 	}
 
-	if err := c.scaleDownDeployments(replicas, desiredDeployments); err != nil {
-		return errors.Wrap(err, "failed to scale down mds deployments")
+	if err := client.FailAllStandbyReplayMDS(c.context, c.clusterInfo, c.fs.Name); err != nil {
```

Comment:
I think it would be good to also add more comment here.
```
// In Pacific, standby-replay daemons are stopped automatically. Older versions of Ceph require us to stop these daemons manually.
```